File: go-react-dpop/dpopUtil.go
package main

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"math/big"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt/v5"
	"github.com/patrickmn/go-cache"
)

// Using go-cache for this demo, would use Redis/database for production
var jtiCache = cache.New(5*time.Minute, 5*time.Minute)

func ValidateDPoPMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {

		// Validate DPoP proof before issuing token
		dpopProof := c.GetHeader("DPoP")
		if dpopProof == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "DPoP proof required"})
			c.Abort()
			return
		}

		url := "http://" + c.Request.Host + c.Request.URL.Path
		method := c.Request.Method

		jkt, err := validateDPoPProof(dpopProof, method, url)
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": err.Error()})
			c.Abort()
			return
		}

		c.Set("dpop_jkt", jkt)
		c.Next()
	}
}

type DPoPClaims struct {
	HttpMethod    string `json:"htm"`
	HttpTargetUri string `json:"htu"`
	JwtId         string `json:"jti"`
	jwt.RegisteredClaims
}

func validateDPoPProof(proof string, requestMethod string, requestUrl string) (string, error) {

	fmt.Println("Validating DPoP proof:", proof, "with method:", requestMethod, "and url:", requestUrl)

	claims := &DPoPClaims{}
	token, _, err := jwt.NewParser().ParseUnverified(proof, claims)
	if err != nil {
		return "", err
	}

	// Validate header
	if token.Header["typ"] != "dpop+jwt" {
		return "", errors.New("Invalid typ, expected 'dpop+jwt'")
	}
	if token.Header["alg"] != "ES256" {
		return "", errors.New("Invalid alg, expected 'ES256'")
	}
	jwkRaw, ok := token.Header["jwk"].(map[string]interface{})
	if !ok {
		return "", errors.New("Missing or invalid JWK in header")
	}
	publicKey, err := jwkToECDSAPublicKey(jwkRaw)
	if err != nil {
		return "", fmt.Errorf("invalid JWK: %w", err)
	}

	// Validate claims
	if claims.HttpMethod != requestMethod {
		return "", fmt.Errorf("Method mismatch: expected %s, got %s", requestMethod, claims.HttpMethod)
	}
	if claims.HttpTargetUri != requestUrl {
		return "", fmt.Errorf("Url mismatch: expected %s, got %s", requestUrl, claims.HttpTargetUri)
	}
	if time.Now().Unix()-claims.IssuedAt.Unix() > 60 {
		return "", errors.New("DPoP proof expired")
	}
	if err := jtiCache.Add(claims.JwtId, true, 1*time.Minute); err != nil {
		return "", errors.New("JTI already used")
	}

	// Verify signature
	_, err = jwt.Parse(proof, func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodECDSA); !ok {
			return nil, fmt.Errorf("Unexpected signing method: %v", t.Header["alg"])
		}
		return publicKey, nil
	})
	if err != nil {
		return "", fmt.Errorf("signature verification failed: %w", err)
	}

	// Calculate and return JKT
	return createJKT(jwkRaw), nil
}

func jwkToECDSAPublicKey(jwk map[string]interface{}) (*ecdsa.PublicKey, error) {
	if jwk["kty"] != "EC" || jwk["crv"] != "P-256" {
		return nil, errors.New("Unsupported key type, expected EC P-256")
	}

	xBytes, err := base64.URLEncoding.DecodeString(jwk["x"].(string))
	if err != nil {
		return nil, errors.New("Failed to decode x coordinate of key")
	}
	yBytes, err := base64.URLEncoding.DecodeString(jwk["y"].(string))
	if err != nil {
		return nil, errors.New("Failed to decode y coordinate of key")
	}

	return &ecdsa.PublicKey{
		Curve: elliptic.P256(),
		X:     new(big.Int).SetBytes(xBytes),
		Y:     new(big.Int).SetBytes(yBytes),
	}, nil
}

func createJKT(jwk map[string]interface{}) string {
	hash := sha256.Sum256([]byte(fmt.Sprintf(`{"crv":"%s","kty":"%s","x":"%s","y":"%s"}`,
		jwk["crv"], jwk["kty"], jwk["x"], jwk["y"])))
	return base64.RawURLEncoding.EncodeToString(hash[:])
}

File: go-react-dpop/src/dpopUtil.ts
import * as indexeddb from "./indexeddb";
import * as jose from "jose";

export async function generateAndSaveKeyPair(): Promise<void> {
  console.log("=== Generating keypair...");
  const keyPair = await crypto.subtle.generateKey(
    {
      name: "ECDSA",
      namedCurve: "P-256",
    },
    false,
    ["sign", "verify"]
  );
  const publicKeyJwk = await crypto.subtle.exportKey("jwk", keyPair.publicKey);
  console.log("Generated public key:", publicKeyJwk);
  try {
    console.log("Generated private key:");
    const privateKeyJwk = await crypto.subtle.exportKey(
      "jwk",
      keyPair.privateKey
    );
  } catch (e) {
    console.log(e);
  }
  indexeddb.saveKeyPair(keyPair);
}

export async function createDPoPProof(
  method: string,
  url: string
): Promise<string> {
  console.log(`=== Creating DPoP proof for method ${method} and url ${url}...`);

  const keyPair = await indexeddb.getKeyPair();
  if (!keyPair) {
    throw new Error("No key pair found in IndexedDB. Please generate a key pair first.");
  }

  const publicKeyJwk = await crypto.subtle.exportKey("jwk", keyPair.publicKey);

  const dpopProof = await new jose.SignJWT(
    // Body of the DPoP proof JWT
    {
      htm: method,
      htu: url,
      jti: crypto.randomUUID(),
      iat: Math.floor(Date.now() / 1000),
    }
  )
    .setProtectedHeader(
      // Header of the DPoP proof JWT
      {
        typ: "dpop+jwt",
        alg: "ES256",
        jwk: publicKeyJwk,
      }
    )
    .sign(keyPair.privateKey);

  console.log("Created DPoP proof:", formatJwt(dpopProof));
  return dpopProof;
}

function formatJwt(jwt: string): string {
  const [headerB64, payloadB64, signature] = jwt.split(".");

  const decodeBase64Url = (str: string) => {
    const base64 = str.replace(/-/g, "+").replace(/_/g, "/");
    return JSON.parse(atob(base64));
  };

  const header = JSON.stringify(decodeBase64Url(headerB64), null, 2);
  const payload = JSON.stringify(decodeBase64Url(payloadB64), null, 2);

  return `
  ==================
  Raw: ${jwt}
  
  Header:
  ${header}
  
  Payload:
  ${payload}
  
  Signature: ${signature}
  ==================
  `;
}
